scipy/stats/_distn_infrastructure.py

```diff
@@ -2853,6 +2854,108 @@ def entropy(pk, qk=None, base=None, axis=0):
     return S
 
 
+def differential_entropy(
+    values: np.ndarray,
+    *,
+    window_length: int = 1,
+    base: Optional[float] = None,
+    axis: int = 0,
+) -> np.ndarray:
+    r"""Given a sample of a distribution, calculate the differential entropy.
+
+    This routine uses the Vasicek estimator of the differential entropy. Given
+    a sorted random sample :math:`X_1, \ldots X_n` this is defined as:
+
+    .. math::
+        \frac{1}{n}\sum_1^n \log\left[ \frac{n}{2m} (X_{i+m} - X_{i-m}) \right]
+
+    where :math:`m` is the window length parameter, :math:`X_{i} = X_1` if
+    :math:`i < 1` and :math:`X_{i} = X_n` if :math:`i > n`.
+
+    Parameters
+    ----------
+    values : sequence
+        Samples of the (continuous) distribution.
+    window_length : int, optional
+        Window length for computing Vasicek estimate. Must be an integer
+        between 1 and half of the sample size.
+    base : float, optional
+        The logarithmic base to use, defaults to ``e`` (natural logarithm).
+    axis: int, optional
+        The axis along which the differential entropy is calculated.
+        Default is 0.
+
+    Returns
+    -------
+    entropy : float
+        The calculated differential entropy.
+
+    Examples
+    --------
+
+    >>> from scipy.stats import differential_entropy, norm
+
+    Entropy of a standard normal distribution:
+
+    >>> random_state = np.random.RandomState(0)
+    >>> values = random_state.standard_normal(100)
+    >>> differential_entropy(values)
+    1.1220441777259473
+
+    Compare with the true entropy:
+
+    >>> float(norm.entropy())
+    1.4189385332046727
+
+    Increasing the window length can improve the accuracy:
+
+    >>> differential_entropy(values, window_length=8)
+    1.3494014875503249
+
```

Comment:
I was hoping that the estimate would improve as the number of sample values increased, but that doesn't appear to be the case unless the window size is increased. Are there any heuristics for choosing the window size that work well in practice?

Reply:
I have clarified in the text the convergence requirements (taken from the article). I also have cited the article in the examples, as it claims that for a given n smoother densities benefit from larger window lengths.

Reply:
Interesting. We will let the reader interpret that to choose the window size for now.

In the future, the "Notes" section might be a better place for these details. (See [numpydoc style guide](https://numpydoc.readthedocs.io/en/latest/format.html) section 13.)

Reply:
On second thought, what do you think about the heuristic [here](https://link.springer.com/content/pdf/10.1007/s40745-015-0045-9.pdf)?
![image](https://user-images.githubusercontent.com/6570539/110227716-2977e800-7eb0-11eb-80ee-10eda2b02157.png)

(notation is the same as here)

Reply:
Seems useful.

![image](https://user-images.githubusercontent.com/6570539/110228495-78c11700-7eb6-11eb-9c1b-79a25976cbdc.png)

Reply:
> On second thought, what do you think about the heuristic [here](https://link.springer.com/content/pdf/10.1007/s40745-015-0045-9.pdf)?

That heuristic formula behaves well asymptotically but I do not see that they mention any theoretical advantage for a fixed n in the papers proposing it. Also, I guess that your experiment is done using a standard normal distribution, but maybe selecting m in that way is not a good choice for other distributions.

However, given that this heuristic is widely used (and better that a fixed value of m), I think that it might be a better choice for the default parameter, so I will change it.

I wonder if it makes sense in the general case to allow that parameter to be a callable that depends on the sample size, or if we should leave that computation to the user in order to keep the interface simple.

Reply:
> allow that parameter to be a callable that depends on the sample size

Good idea, and it's not inherently complicated, but if the user is going to write the candle, I'm not sure if there is an advantage passing it as an argument over calling it, then passing the output as an argument. So might as well leave it simple.

It would be great to see a `method` parameter, as that reference gives several other estimation methods that would be very easy to add now that the framework is here. That can (and should) be in a separate PR to avoid holding this one up. I can take care of it if you're not interested, though.

Reply:
The parameter `window_length` now is `None` by default and uses the proposed heuristic in that case. It has been documented and the original paper proposing it (not the one you said, but the one referenced in that work) has been added to the references.

I do not think that I will add more heuristics after this is merged, so feel free to continue improving this function afterwards.